outlines/fsm/fsm.py

```diff
@@ -114,27 +115,48 @@ def __init__(
         regex_string: str,
         tokenizer: "Tokenizer",
     ):
-        regex_pattern = interegular.parse_pattern(regex_string)
-        regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
+        def func_cache_key_args(
+            regex_string: str, tokenizer: "Tokenizer"
+        ) -> Tuple[str, list]:
+            """Return the values that will be used to create the cache key of create_states_mapping"""
+            cacheable_vocabulary = sorted(tokenizer.vocabulary.values())
+            return (regex_string, cacheable_vocabulary)
+
+        @cache(func_cache_key_args)
+        def create_states_mapping(
+            regex_string: str, tokenizer: "Tokenizer"
+        ) -> Tuple[dict, set, set]:
+            """Create the variables related the mapping between stzates and tokens
+            The parameters of the function are used for caching purpose
+            """
+            regex_pattern = interegular.parse_pattern(regex_string)
+            regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
+            (
+                states_to_token_maps,
+                empty_token_ids,
+            ) = create_fsm_index_tokenizer(regex_fsm, tokenizer)
+
+            # We make sure that it is possible to generate strings in the language
+            # of the regular expression with the tokens present in the model's
+            # vocabulary.
+            if not any(
+                regex_fsm.finals.intersection(v.values())
+                for v in states_to_token_maps.values()
+            ):
+                raise ValueError(
+                    "The vocabulary does not allow us to build a sequence that matches the input regex"
+                )
+
+            final_states = regex_fsm.finals | {
+                -1
```

Comment:
What does `-1` indicate in the context of a `RegexFSM` state?

Reply:
Everything that is a final state not defined by the regex. This was used to stop generation when `max_token` was found but I'm not sure that's relevant anymore.